README.md
# prediction

a package for predicting if cancer is treatable or not

## Installation

```bash
$ pip install dsci-prediction
```

## Usage

- TODO

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`prediction` was created by Clichy Bazenga. It is licensed under the terms of the MIT license.

## Credits

`prediction` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (04/04/2022)

- First release of `prediction`!

tests/test_mean.py
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from prediction.prediction import mean_cross_val_scores


class Data:

    def __init__(self):
        col_names = ["id", "clump", "unif_size",
                     "unif_shape", "adhesion", "epi_size", "nuclei",
                     "chromatin", "nucleoli", "mitoses", "class"]
        
        dataset = pd.read_csv("data/raw/breast_cancer.txt", names=col_names, sep=",")
    
        dataset = dataset[(dataset != '?').all(axis=1)]
        dataset['nuclei'] = dataset['nuclei'].astype(int)
        dataset = dataset.drop(columns=["id"])
        dataset['class'] = dataset['class'].replace([2], 0)
        dataset['class'] = dataset['class'].replace([4], 1)
        train_df, test_df = train_test_split(dataset,
                                             test_size=0.3, random_state=123)
        self.X_train = train_df.drop(columns=["class"])
        self.X_test = test_df.drop(columns=["class"])
        self.y_train = train_df["class"]
        self.y_test = test_df["class"]


def test_mean_cross_val_correct_simple():
    """mean_cross_val_scores returns correct mean
    of cross validation with correct types of inputs""" 
    """test if it is a series"""
    """test if num of elements are as expected"""
    """test if elements bigger than 0 and smaller than 1"""
    dataset = Data()
    scale = StandardScaler()
    pipe_knn = make_pipeline(scale,
                             KNeighborsClassifier(n_neighbors=5))
    scoring = ['accuracy']
    result = mean_cross_val_scores(
        pipe_knn, dataset.X_train, dataset.y_train,
        return_train_score=True, scoring=scoring)
    assert all(result) <= 1
    assert all(result) >= 0
    assert result.count() == 4
    assert isinstance(result, pd.core.series.Series)


def test_mean_cross_val_correct():
    """mean_cross_val_scores returns correct mean
    of cross validation with correct types of inputs""" 
    """test if it is a series"""
    """test if num of elements are as expected"""
    """test if elements bigger than 0 and smaller than 1"""
    dataset = Data()
    scale = StandardScaler()
    pipe_knn = make_pipeline(scale,
                             DecisionTreeClassifier(random_state=123))
    scoring = ['accuracy', "f1", "recall", "precision"]
    result = mean_cross_val_scores(
        pipe_knn, dataset.X_train, dataset.y_train,
        return_train_score=True, scoring=scoring)
    assert all(result) <= 1
    assert all(result) >= 0
    assert result.count() == 10
    assert isinstance(result, pd.core.series.Series)
    

tests/test_plot_hist.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from src.plot_hist import plot_hist_overlay
import pytest

# simpler dataframe
df0 = pd.DataFrame(np.linspace(2,10,20), columns=["x1"])
df1 = pd.DataFrame(np.linspace(4,10,20), columns=["x1"])

# more complex (edge_cases) dataframe
square_dict0 = {"feat1" : np.linspace(2,10,20),
               "feat2" : np.logspace(2,10,20),
               "feat3" : np.linspace(2,10,20),
               "feat4" : np.linspace(2,10,20)}

square_dict1 = {"feat1" : np.linspace(4,10,20),
               "feat2" : np.logspace(4,10,20),
               "feat3" : np.linspace(4,10,20),
               "feat4" : np.linspace(4,10,20)}

sqr_df0 = pd.DataFrame(square_dict0)
sqr_df1 = pd.DataFrame(square_dict1)
feats = ["feat1", "feat2", "feat3", "feat4"]

dict0 = {"feat1" : np.linspace(2,10,20),
        "feat2" : np.logspace(2,10,20),
        "feat3" : np.linspace(2,10,20)}
    
dict1 = {"feat1" : np.linspace(4,10,20),
        "feat2" : np.logspace(4,10,20),
        "feat3" : np.linspace(4,10,20)}

notsqr_df0 = pd.DataFrame(dict0)
notsqr_df1 = pd.DataFrame(dict1)
feats2 = ["feat1", "feat2", "feat3"]

# initial objects to test with
labels = ["0 - negative","1 - positive"]
fig, axe = plot_hist_overlay(df0, df1, ["x1"], labels=labels)
fig_test, ax_test = plt.subplots()

def test_return_type():
    """
    Test for the correct return type of the function is an Axes object
    """
    assert type(axe) == type(ax_test)
    assert type(fig) == type(fig_test)

def test_incorrect_parameter_type():
    """
    Test if a TypeError is raise if any input parameter does not have the correct type
    as required in the function documentation
    """
    with pytest.raises(TypeError):
        plot_hist_overlay("df0", df1, labels=labels, columns=feats)

    with pytest.raises(TypeError):
        plot_hist_overlay(df0, "df1", labels=labels, columns=feats)
    
    with pytest.raises(TypeError):
        plot_hist_overlay("df0", "df1", labels="label", columns=feats)
    
    with pytest.raises(TypeError):
        plot_hist_overlay(df0, df1, labels=labels, columns="col")
    
    with pytest.raises(TypeError):
        plot_hist_overlay("df0", df1, labels=labels, columns=["x1"],fig_no=1) 

def test_genearal_readability():
    """
    Test for the correct label for X,y axis, legend and title
    """
    assert axe.get_xlabel() == "X1"
    assert axe.get_ylabel() == "Count"
    assert axe.get_legend().get_texts()[0].get_text() == labels[0]
    assert axe.get_legend().get_texts()[1].get_text() == labels[1]
    
    # default figure number fig_no = 1
    assert axe.get_title() == "Figure 1.1: Histogram of X1 for each target class label"
    
    # figure number supplied by fig_no
    fig2,axe2 = plot_hist_overlay(df0, df1, ["x1"], fig_no="2", labels=labels, ec="white")
    
    assert axe2.get_title() == "Figure 2.1: Histogram of X1 for each target class label"
    plt.cla()

def test_edge_case_1(): 
    """
    Test to see if the function correctly plot for dataframe with only 1 column
    """
    fig3,axe3 = plot_hist_overlay(df0, df1, ["x1"], fig_no="2", labels=labels, ec="white")
    
    assert type(axe3) == type(ax_test)
    assert axe3.get_legend().get_texts()[0].get_text() == labels[0]
    assert axe3.get_legend().get_texts()[1].get_text() == labels[1]
    assert axe3.get_title() == "Figure 2.1: Histogram of X1 for each target class label"

def test_edge_case_square():
    """
    Test functionality when the whole figure is a square of subplots, e.g: dimension of a square (2,2)
    with the number of features is perfect squares e.g: 4 features can be arrange into a 2x2 square figure
    """
    sqr_fig, sqr_axe = plot_hist_overlay(sqr_df0, sqr_df1, columns=feats, labels=labels, fig_no="3")
    
    # basic functionalities
    assert type(sqr_axe) == type(ax_test)
    assert sqr_axe.get_xlabel() == "Feat4"
    assert sqr_axe.get_ylabel() == "Count"
    assert sqr_axe.get_legend().get_texts()[0].get_text() == labels[0]
    assert sqr_axe.get_legend().get_texts()[1].get_text() == labels[1]

    # test correct dimension that is automatically calculated by our function
    sqr_fig2, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2)
    assert sqr_axe.get_gridspec().ncols == ax4.get_gridspec().ncols
    assert sqr_axe.get_gridspec().ncols == 2
    assert sqr_axe.get_gridspec().nrows == ax4.get_gridspec().nrows
    assert sqr_axe.get_gridspec().nrows == 2
    assert sqr_axe.get_title() == "Figure 3.4: Histogram of Feat4 for each target class label"

def test_edge_case_not_square():
    """
    Test functionality when the figure is a square but the number of features are not quite perfect square
    Will also test if we can render an empty subplot
    """
    notsqr_fig, notsqr_axe = plot_hist_overlay(notsqr_df0, notsqr_df1, columns=feats2, labels=labels, fig_no="4")

    # basic functionalities
    assert type(notsqr_axe) == type(ax_test)
    assert notsqr_axe.get_xlabel() == "Feat3"
    assert notsqr_axe.get_ylabel() == "Count"
    assert notsqr_axe.get_legend().get_texts()[0].get_text() == labels[0]
    assert notsqr_axe.get_legend().get_texts()[1].get_text() == labels[1]

    # test correct dimension that is automatically calculated by our function
    # in this case 3 features will also be plotted in a square grid 2x2, but the last subplot is just empty
    sqr_fig2, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2,2)
    assert notsqr_axe.get_gridspec().ncols == ax3.get_gridspec().ncols
    assert notsqr_axe.get_gridspec().ncols == 2
    assert notsqr_axe.get_gridspec().nrows == ax3.get_gridspec().nrows
    assert notsqr_axe.get_gridspec().nrows == 2
    assert notsqr_axe.get_title() == "Figure 4.3: Histogram of Feat3 for each target class label" 


tests/test_plot_boxplot.py
from unicodedata import numeric
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
from dsci_prediction.dsci_prediction import *

test_df = pd.DataFrame({'age':['25','48','30'], 'height': ['185','192','187'],'weight':['85','93','90'], 
'class':['0','1','1'] })
num_df=test_df.apply(pd.to_numeric)
var_names=num_df.head()
number_of_rows=3
number_of_columns=1
test_case = boxplot_plotting(number_of_rows,number_of_columns,0.5,0.5,var_names,num_df,3)
b=mpl.text.Text()
comparison_var=5

def test_return_type():
    #Test for the correct return type of function:
    assert type(test_case) == type(b)

def test_dataframe_type_values():
    #Tests to see if the values of each column are numeric in order to be able to plot them
    for i in range (len(var_names)):
        assert type(i)==type(comparison_var)

def test_product_consistency():
    #Tests to see if the number of boxplots created will match the number of variables involved. This is 
    #to avoid extra unuseful boxplots or not enough boxplots to show all variables interacting with the class values
    assert number_of_columns * number_of_rows == len(var_names)


src/DSCI_prediction/DSCI_prediction.py
import pandas as pd
import numpy as np
import argparse

import matplotlib.pyplot as plt

import seaborn as sns


import sklearn


def EDA_plot(train_df, hist_output, boxplot_output):
    """The purpose of this function is to plot the training data, with their given output class, in both histogram and boxplots. Afterwards, the plots are saved for further use. """
    train_df = pd.read_csv(str(train_df))
    X_train = train_df.drop(columns=["class"])
    numeric_looking_columns = X_train.select_dtypes(
        include=np.number).columns.tolist()
    benign_cases = train_df[train_df["class"] == 0]
    malignant_cases = train_df[train_df["class"] == 1]
    #plot histogram
    fig,ax = plot_hist_overlay(df0=benign_cases, df1=malignant_cases,
                 columns=numeric_looking_columns, labels=["0 - benign", "1 - malignant"],
                 fig_no="1")
    fig.savefig(str(hist_output), facecolor="white")
    #plot boxplot 
    fig2 = boxplot_plotting(3, 3, 20, 25, numeric_looking_columns, train_df, 2)
    fig2.savefig(str(boxplot_output), facecolor="white")

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Plots EDA")
    parser.add_argument("train_df", help="Path to train_df")
    parser.add_argument("hist_output", help="Path to histogram output")
    parser.add_argument("boxplot_output", help="Path to boxplot output")
    args = parser.parse_args()
    EDA_plot(args.train_df, args.hist_output, args.boxplot_output)
    
    
    
def plot_hist_overlay(df0, df1, columns, labels, fig_no="1",alpha=0.7, bins=5, **kwargs):
    """
    A function that plot multiple histogram for a target
    classification label against each numerical features.
    The resulting histograms will be a grid layout contained in
    one single Figure object
    PARAMETERS:
    -------
    df0:
        A pandas DataFrame that is corresponded to the label 0
    df1:
        A pandas DataFrame that is corresponded to the label 1
    columns:
        A list of column name
    labels: 
        A list of label for each of the histograms for each label
    fig_no: optional, default="1"
        A string denoting the figure number, in the case of multiple figures
    alpha: optional, default=0.7
        A float denotes the alpha value for the matplotlib hist function
    bin: optional, default=5
        An int denotes the number of bins for the matplotlib hist function
    **kwargs:
        Other parameters for the plotting function
    REQUISITES: 
    target label are binary i.e 0 or 1, negative or positive
    -------
    RETURNS:
    -------
    A matplotlib.figure.Figure object
    Examples:
    -------
    benign_cases = train_df[train_df["class"] == 0]   # df0             
    malignant_cases = train_df[train_df["class"] == 1] # df1
    plot_hist_overlay(benign_cases, malignant_cases,["unif_size"], labels=["0 - benign", "1 - malignant"]
    
    """
    # These are legacy codes are comment out in case we need to reuse in the future
    # column_name = column.title().replace("_", " ")
    # fig, ax = plt.subplots()
    # ax.hist(df0[column], alpha=alpha, bins=bins, label=labels[0], **kwargs, figure=fig)
    # ax.hist(df1[column], alpha=alpha, bins=bins, label=labels[1], **kwargs, figure=fig)
    # ax.legend(loc="upper right")
    # ax.set_xlabel(column_name)
    # ax.set_ylabel("Count")
    # ax.set_title(f"Figure {fig_no}: Histogram of {column_name} for each target class label")
    # return ax

    if not isinstance(df0, (pd.core.series.Series,
                                pd.core.frame.DataFrame, np.ndarray)):
        raise TypeError("'df0' should be of type numpy.array or pandas.Dataframe")
    if not isinstance(df1, (pd.core.series.Series,
                                pd.core.frame.DataFrame, np.ndarray)):
        raise TypeError("'df1' should be of type numpy.array or pandas.Dataframe")
    if not isinstance(columns, list):
        raise TypeError("'columns' should be of type list")
    if not isinstance(labels, list):
        raise TypeError("'labels' should be of type list")
    if not isinstance(fig_no, str):
        raise TypeError("'fig_no' should be of 'str'")

    ## other parameters are supplied into the matplotlib functions

    # To automatically calculating the size of dimension of the figures (Square shape)
    size = len(columns)
    dim = np.ceil(np.sqrt([size])).astype(int)[0]
    fig = plt.figure(1, figsize=(22,22))

    for idx, x in enumerate(columns):
        subplot=plt.subplot(dim, dim, idx+1)
        col_name = x.title().replace("_", " ")
        subplot.hist(df0[x], alpha=alpha, bins=bins, label=labels[0], **kwargs)
        subplot.hist(df1[x], alpha=alpha, bins=bins, label=labels[1], **kwargs)
        subplot.legend(loc="upper right")
        subplot.set_xlabel(col_name, fontsize=14)
        subplot.set_ylabel("Count", fontsize=14)
        subplot.set_title(f"Figure {fig_no}.{idx+1}: Histogram of {col_name} for each target class label", 
                          fontsize=14)

    return (fig, subplot)



def boxplot_plotting (num_rows,num_columns,width,height,variables,datafr,number):
    """
    A function which returns a given number of boxplots for different target  against each numerical feature. The returning objects are seaborn.boxplot types. 
    
    -------------------
    PARAMETERS:
    A dataframe containing the variables and their correspondent labels
    Variables: A list of each variable's name
    num_rows and num_columns: An integer and positive number for both num_rows and num_columns for the
    boxplot fig "canvas" object where our boxplots will go,
    width: A positive width measure 
    length: A positive length measure 
    A binary class label 
    A column array for managing variable names
    A training dataframe object
    Integer positive number for correct ordering  of graphs 
    -------------------
    REQUISITES:
    The target labels ("class label") must be within the data frame 
    The multiplication between num_rows and num_columns must return be equal to num_variables.
    It is possible for num_rows & num_columns to be values that when multiplied don't equal the "variables" numeric value,
    but that will create more boxplots which will be empty. 
    

    --------------------
    RETURNS:
    It returns a fixed number "num_variables" of boxplot objects. Each Boxplot represents both Target Class
    Labels according to a given Variable

    --------------------
    Examples

    datafr=train_df
    --------
    boxplot_plotting (3,3,20,25,numeric_column,datafr,number)
    """
    fig,ax= plt.subplots(num_rows,num_columns,figsize=(width,height))
    for idx, (var,subplot) in enumerate(zip(variables,ax.flatten())):
        a = sns.boxplot(x='class',y=var,data=datafr,ax=subplot).set_title(f"Figure {number}.{idx}: Boxplot of {var} for each target class label")
    return fig



def tuned_para_table(search, X_train, y_train):
    """
    A function which returns a panda dataframe of tuned hyperparameters
    and its best score given GridSearchCV object fitted X_train and y_train
    -------------------
    PARAMETERS:
    search: A sklearn.model_selection._search.GridSearchCV that has been
    specified estimator, param_grid, **kwargs
    X_train : numpy array or pandas DataFrame/Series
        X in the training data
    y_train : numpy array or pandas DataFrame/Series
        y in the training data
    --------------------
    REQUISITES:
    X_train, y_train must at least n_splits (specified in cv in search)
    observations for each target class.
    search must be GridSearchCV object that is clearly specified with
    estimator, param_grid, cv, and so on.
    --------------------
    RETURNS:
    Returns a pandas.core.frame.DataFrame object that specifies
    the tuned hyperaparameters and the best score produced by GridSearchCV
    --------------------
    Examples

    search = GridSearchCV(KNeighborsClassifier(),
                      param_grid={'kneighborsclassifier__n_neighbors':
                      range(1, 10)},
                      cv=10, 
                      n_jobs=-1,  
                      scoring="recall", 
                      return_train_score=True)
    --------
    tuned_para_table(search, X_train, y_train)
    """
    if not isinstance(search, sklearn.model_selection._search.GridSearchCV):
        raise TypeError("'search' should be of type GridSearchCV")
    if not isinstance(X_train, (pd.core.series.Series,
                                pd.core.frame.DataFrame, np.ndarray)):
        raise TypeError("'X_train' should be of type np.array or pd.Dataframe")
    if not isinstance(y_train, (pd.core.series.Series,
                                pd.core.frame.DataFrame, np.ndarray)):
        raise TypeError("'y_train' should be of type np.array or pd.Dataframe")
    search.fit(X_train, y_train)
    best_score = search.best_score_.astype(type('float', (float,), {}))
    tuned_para = pd.DataFrame.from_dict(search.best_params_, orient='index')
    tuned_para = tuned_para.rename(columns = {0 : "Value"})
    tuned_para = tuned_para.T
    tuned_para['best_score'] = best_score
    return tuned_para


src/prediction/prediction.py

from sklearn.model_selection import train_test_split
import pandas as pd
import argparse
from sklearn.model_selection import cross_validate

import numpy as np
import matplotlib.pyplot as plt


from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    ConfusionMatrixDisplay
)

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.compose import make_column_transformer

import seaborn as sns

def load_data(input_path, output_path):
    
    '''
    Insert comment
    
    '''
    col_names = ["id", "clump", "unif_size", "unif_shape", "adhesion", "epi_size", "nuclei", "chromatin", "nucleoli", "mitoses", "class"]
    dataset = pd.read_csv(str(input_path), names=col_names, sep=",")
    
    return dataset.to_csv(str(output_path), index=False)


def clean_data(input_path, output_path_train, output_path_test):
    '''
    Insert comment
    '''
    
    #cleaning data
    
    df = pd.read_csv(str(input_path))
    df = df[(df != '?').all(axis=1)]
    df['nuclei'] = df['nuclei'].astype(int)
    df = df.drop(columns=["id"])
    
    #replace 2 -> 0 & 4 -> 1 in target class 
    df['class'] = df['class'].replace([2],0)
    df['class'] = df['class'].replace([4],1) 
    
    #split train/test data
    train_df, test_df = train_test_split(df, test_size=0.3, random_state=123)
    train_df.to_csv(str(output_path_train), index=False)
    test_df.to_csv(str(output_path_test), index=False)
    
    
def EDA_plot(train_df, hist_output, boxplot_output):
    
    '''
    Insert comment
    '''
    
    train_df = pd.read_csv(str(train_df))
    X_train = train_df.drop(columns=["class"])
    numeric_looking_columns = X_train.select_dtypes(
        include=np.number).columns.tolist()
    benign_cases = train_df[train_df["class"] == 0]
    malignant_cases = train_df[train_df["class"] == 1]
    
    #plot histogram
    fig = plot_hist_overlay(df0=benign_cases, df1=malignant_cases,
                            columns=numeric_looking_columns, 
                            labels=["0 - benign", "1 -malignant"],
                            fig_no="1")
    fig.savefig(str(hist_output), facecolor="white")
    
    #plot boxplot 
    fig2 = boxplot_plotting(3, 3, 20, 25, numeric_looking_columns, train_df, 2)
    fig2.savefig(str(boxplot_output), facecolor="white")
    
    scoring = [
        "accuracy",
        "f1",
        "recall",
        "precision",]
    
    results = {}
    
    
def build_test_model(train_df, test_df, cross_val_output, tuned_para_output,
                     classification_output, confusion_matrix_output):
    '''
    Insert comment
    '''
    
    np.random.seed(123)
    train_df = pd.read_csv(str(train_df))
    test_df = pd.read_csv(str(test_df))
    X_train = train_df.drop(columns=["class"])
    X_test = test_df.drop(columns=["class"])
    y_train = train_df["class"]
    y_test = test_df["class"]
    numeric_looking_columns = X_train.select_dtypes(include=np.number).columns.tolist()
    numeric_transformer = StandardScaler()
    ct = make_column_transformer((numeric_transformer, numeric_looking_columns))
    pipe_knn = make_pipeline(ct, KNeighborsClassifier(n_neighbors=5))
    pipe_dt = make_pipeline(ct, DecisionTreeClassifier(random_state=123))
    pipe_reg = make_pipeline(ct, LogisticRegression(max_iter=100000))
    classifiers = {
        "kNN": pipe_knn,
        "Decision Tree": pipe_dt,
        "Logistic Regression" : pipe_reg}
    
    #cross_val_scores_for_models
    for (name, model) in classifiers.items():
        results[name] = mean_cross_val_scores(
            model,
            X_train,
            y_train,
            return_train_score=True,
            scoring = scoring)
        cross_val_table = pd.DataFrame(results).T
        cross_val_table.to_csv(str(cross_val_output))
        
        #tune hyperparameters 
        np.random.seed(123)
        search = GridSearchCV(pipe_knn,
                              param_grid={'kneighborsclassifier__n_neighbors': range(1,50),
                                          'kneighborsclassifier__weights': ['uniform', 'distance']},
                              cv=10,
                              n_jobs=-1,
                              scoring="recall",
                              return_train_score=True)
        search.fit(X_train, y_train)
        best_score = search.best_score_.astype(type('float', (float,), {}))
        tuned_para = pd.DataFrame.from_dict(search.best_params_, orient='index')
        tuned_para = tuned_para.rename(columns = {0:"Value"})
        tuned_para = tuned_para.T
        tuned_para['knn_best_score'] = best_score
        tuned_para.to_csv(str(tuned_para_output))
        
        #model on test set 
        pipe_knn_tuned = make_pipeline(ct,KNeighborsClassifier(
            n_neighbors=search.best_params_['kneighborsclassifier__n_neighbors'],
            weights=search.best_params_['kneighborsclassifier__weights']))
        pipe_knn_tuned.fit(X_train, y_train)
        
        #classification report 
        report = classification_report(y_test, pipe_knn_tuned.predict(X_test),
                                       output_dict=True, target_names=["benign", "malignant"])
        report = pd.DataFrame(report).transpose()
        report.to_csv(str(classification_output))
        
        #confusion matrix 
        cm = confusion_matrix(y_test, pipe_knn_tuned.predict(X_test), labels=pipe_knn_tuned.classes_)
        disp = ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=pipe_knn_tuned.classes_)
        disp.plot()
        plt.savefig(str(confusion_matrix_output))
        from sklearn.model_selection import cross_validate
        

def mean_cross_val_scores(model, X_train, y_train, **kwargs):
    """
    Returns mean of cross validation of given model, X_train, y_train 
    Parameters
    -----------
    model :
        scikit-learn model
    X_train : numpy array or pandas DataFrame
        X in the training data
    y_train : numpy array or pandas DataFrame
        y in the training data
    Returns
    -----------
        pandas Series with mean scores from cross_validation
    """
    scores = cross_validate(model, X_train, y_train, **kwargs)
    mean_scores = pd.DataFrame(scores).mean()
    out_col = []

    for i in range(len(mean_scores)):
        out_col.append(mean_scores[i])

    return pd.Series(data=out_col, index=mean_scores.index)
    
    
def boxplot_plotting (num_rows,num_columns,width,height,variables,datafr,number):
    
    """
    A function which returns a given number of boxplots for different target  
    against each numerical feature. The returning objects are seaborn.boxplot types. 
    -------------------
    PARAMETERS:
    A dataframe containing the variables and their correspondent labels
    Variables: A list of each variable's name
    num_rows and num_columns: An integer and positive number for both num_rows and num_columns for the
    boxplot fig "canvas" object where our boxplots will go,
    width: A positive width measure 
    length: A positive length measure 
    A binary class label 
    A column array for managing variable names
    A training dataframe object
    Integer positive number for correct ordering  of graphs 
    -------------------
    REQUISITES:
    The target labels ("class label") must be within the data frame 
    The multiplication between num_rows and num_columns must return be equal to num_variables.
    It is possible for num_rows & num_columns to be values that when multiplied don't equal the "variables" numeric value,
    but that will create more boxplots which will be empty. 
    
    --------------------
    RETURNS:
    It returns a fixed number "num_variables" of boxplot objects. Each Boxplot represents both Target Class
    Labels according to a given Variable
    --------------------
    Examples
    datafr=train_df
    --------
    boxplot_plotting (3,3,20,25,numeric_column,datafr,number)
    """
    
    fig,ax= plt.subplots(num_rows,num_columns,figsize=(width,height))
    for idx, (var,subplot) in enumerate(zip(variables,ax.flatten())):
        a = sns.boxplot(x='class',y=var,data=datafr,ax=subplot).set_title (f"Figure {number}.{idx}:Boxplot of {var} for each target class label")
    return fig
         
         

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

def plot_hist_overlay(df0, df1, columns, labels, fig_no="1",alpha=0.7, bins=5, **kwargs):
    """
    A function that plot multiple histogram for a target
    classification label against each numerical features.
    The resulting histograms will be a grid layout contained in
    one single Figure object
    REQUIRED: target label are binary i.e 0 or 1, negative or positive
    Parameters
    -------
    df0:
        A pandas DataFrame that is corresponded to the label 0
    df1:
        A pandas DataFrame that is corresponded to the label 1
    columns:
        A list of column name
    labels: 
        A list of label for each of the histograms for each label
    fig_no: optional, default="1"
        A string denoting the figure number, in the case of multiple figures
    alpha: optional, default=0.7
        A float denotes the alpha value for the matplotlib hist function
    bin: optional, default=5
        An int denotes the number of bins for the matplotlib hist function
    **kwargs:
        Other parameters for the plotting function 
    Returns
    -------
    A matplotlib.figure.Figure object
    Examples
    -------
    benign_cases = train_df[train_df["class"] == 0]   # df0             
    malignant_cases = train_df[train_df["class"] == 1] # df1
    plot_hist_overlay(benign_cases, malignant_cases,["unif_size"], labels=["0 - benign", "1 - malignant"]
    
    """
    
    # To automatically calculating the size of dimension of the figures (Square shape)
    size = len(columns)
    dim = np.ceil(np.sqrt([size])).astype(int)[0]
    fig = plt.figure(1, figsize=(22,22))
         
    for idx, x in enumerate(columns):
        subplot=plt.subplot(dim, dim, idx+1)
        col_name = x.title().replace("_", " ")
        subplot.hist(df0[x], alpha=alpha, bins=bins, label=labels[0], **kwargs)
        subplot.hist(df1[x], alpha=alpha, bins=bins, label=labels[1], **kwargs)
        subplot.legend(loc="upper right")
        subplot.set_xlabel(col_name, fontsize=14)
        subplot.set_ylabel("Count", fontsize=14)
        subplot.set_title(f"Figure {fig_no}.{idx+1}: Histogram of {col_name} for each target class label", fontsize=14)
    
    #fig.suptitle(f"Figure {fig_no}: Distribution of the target class for each numeric feature", fontsize=20)
    return fig